### ISSUE TITLE
How to provide supported CUDA compute capabilities and runtime version where needed?

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#33542]
Currently the [`cudaIsEnabled`](https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/CUDAServices/bin/cudaIsEnabled.cpp) uses a hardcoded condition for the CUDA compute capability check.

#33057 adds a need to have the list of supported compute capabilities to `runTheMatrix` as well as the CUDA runtime version (two leading parts if I understood correctly, e.g. `11.2`).

At this point we should look into having a single source of this information that would scale to adding other pieces of information and also to other technologies beyond CUDA.

**Comment by makortel** [comment]
assign core,heterogeneous,pdmv

**Comment by makortel** [comment]
I can think of adding environment variables for these in the CUDA toolfile, but is that the best solution?

**Comment by smuzaffar** [comment]
one other way could be that cuda.spec generates a python script which one can import to get information

**Comment by makortel** [comment]
(to repeat also here) one question for the `runTheMatrix` use case is that different CPU platforms may support different sets of CUDA compute capabilities (e.g. currently ARM does not support Pascal 6.x, whereas x86 and PPC do). When we start to use multiple CPU architectures I'd imagine the job submission to be done on some/any CPU node (e.g. x86 on lxplus).

How to have the correct CUDA compute capability list for ARM jobs? Would the job submission specify the target `SCRAM_ARCH` explicitly? (in which case we could use arch-specific defaults)

**Comment by makortel** [comment]
`cudaIsEnabled` is a C++ program, so how about e.g. a JSON file?

**Comment by smuzaffar** [comment]
json is also possible but then we still need an environment variable to point to it. Can we convert `cudaIsEnabled` in to python?

**Comment by makortel** [comment]
I suppose it could be reimplemented in python by interpreting the output of `cudaComputeCapabilities` that would still communicate to CUDA via C(++) API.

**Comment by makortel** [comment]
Or we remove it altogether and change https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/CUDACore/python/SwitchProducerCUDA.py#L9 to call `cudaComputeCapabilities` directly **if** it would be ok to import that python fragment in CMSSW configuration.

**Comment by fwyzard** [comment]
Note that the compute capabilities that we have in CMSDIST's `cuda-flags.file` are those we compile for, but they are not an exhaustive list of those we can run on.

Currently (on x86 and power) we build for `6.0`, `7.0`, `7.5`.
Most likely the code can then run on any `6.x` and `7.x` GPU (and fail on any `5.x` and `8.x` GPUs).

**Comment by fwyzard** [comment]
I'm confused what you mean with using `cudaComputeCapabilities` ?

`cudaComputeCapabilities` prints the GPUs available on the current system, along with their compute capabilities - but it doesn't know anything about the capabilities that CMSSW supports or not.

**Comment by makortel** [comment]
> Note that the compute capabilities that we have in CMSDIST's `cuda-flags.file` are those we compile for, but they are not an exhaustive list of those we can run on.

Right, so we'd need to provide an explicit list of the compute capabilities where the software can run on (or that's what I understood #33057 would need).

> I'm confused what you mean with using `cudaComputeCapabilities` ?

In python program (be it a separate script or part of the configuration system) execute the program, parse the compute capabilities of the devices from its output, and compare that to the contents of the python fragment generated in `cuda.spec`.

(I think this is still heavily on "what we could do" side rather than "should do")

**Comment by fwyzard** [comment]
> > I'm confused what you mean with using `cudaComputeCapabilities` ?
> 
> In python program (be it a separate script or part of the configuration system) execute the program, parse the compute capabilities of the devices from its output, and compare that to the contents of the python fragment generated in `cuda.spec`.
> 
> (I think this is still heavily on "what we could do" side rather than "should do")

Ah, I see, you mean to decide whether "CUDA is available" locally, independently from the job description/matching.

**Comment by makortel** [comment]
> > > I'm confused what you mean with using `cudaComputeCapabilities` ?
> > 
> > 
> > In python program (be it a separate script or part of the configuration system) execute the program, parse the compute capabilities of the devices from its output, and compare that to the contents of the python fragment generated in `cuda.spec`.
> > (I think this is still heavily on "what we could do" side rather than "should do")
> 
> Ah, I see, you mean to decide whether "CUDA is available" locally, independently from the job description/matching.

Right, this was for `cudaIsEnabled`/ `SwitchProducerCUDA` purpose.

**Comment by fwyzard** [comment]
OK. Then I'd suggest we should autodetect which GPUs are usable based on if we are actually able to use them: https://github.com/cms-sw/cmssw/pull/33561 .

**Comment by makortel** [comment]
I agree, #33561 provides much better way for `cudaIsEnabled` (thanks!). That leaves the need to communicate the compute capabilities where can run and the runtime version only to `runTheMatrix`.

**Comment by fwyzard** [comment]
If we merge https://github.com/cms-sw/cmsdist/pull/6851 , then we will only need to give a "minimum compute capability".

**Comment by fwyzard** [comment]
@makortel @smuzaffar is this still relevant ?
